### PR TITLE
[http2 server] when possible send an END_STREAM flag with HEADERS

### PR DESCRIPTION
**Pull request by deweerdt**
We currently do not send a HEADERS frame with an END_STREAM, relying instead on the ability to send the flag along with a zero sized DATA frame. This causes interoparibility issues with some clients that treat the presence of a DATA frame (empty or not) as semantically different.

This PR makes it so that we set the flag when generating the HEADERS frame, if the information is available at the time (the generator indicated that this is the end of the stream).

### CODE REVIEW

File: lib/http2/stream.c [hunk 1]
```diff
@@ -352,8 +353,12 @@ void finalostream_send(h2o_ostream_t *self, h2o_req_t *req, h2o_sendvec_t *bufs,
             return;
         }
         h2o_probe_log_response(&stream->req, stream->stream_id);
-        if (send_headers(conn, stream) != 0)
+        if (send_headers(conn, stream, is_end_stream) != 0)
             return;
+        if (is_end_stream) {
+            h2o_http2_stream_set_state(conn, stream, H2O_HTTP2_STREAM_STATE_END_STREAM);
+            break;
```
Comment (kazuho) [inline lib/http2/stream.c:360]: By breaking here, we will be calling `h2o_http2_conn_register_for_proceed_callback`. Does not that lead to us sending a zero-sized DATA frame with the END_STREAM flag?

I know that the test does not detect such a behavior but wonder if that is just due to how nghttp works (i.e., close the connection before processing any extraneous frames that might follow).

If that is the case, maybe doing something like `send_refused_stream` called right above might make more sense.

**Review by kazuho — COMMENTED** [review]
Thank you for bringing the issue to our attention.

Reading https://lists.w3.org/Archives/Public/ietf-http-wg/2020OctDec/thread.html#msg40, it seems to me that while people tend to think that our current behavior is acceptable, there are interoperability concerns.

I grudgingly agree that we should better adjust our behavior.

File: lib/http2/stream.c (hunk 1 above)
Comment (deweerdt) [reply lib/http2/stream.c]: Thanks for catching that, i've updated the test to catch the case and fixed the code in 5507e3d

**Comment by deweerdt** [comment]
> Thank you for bringing the issue to our attention.
> 
> Reading https://lists.w3.org/Archives/Public/ietf-http-wg/2020OctDec/thread.html#msg40, it seems to me that while people tend to think that our current behavior is acceptable, there are interoperability concerns.
> 
> I grudgingly agree that we should better adjust our behavior.

Apparently, it's this bit:
> Response-Headers & Trailers-Only are each delivered in a single HTTP2 HEADERS frame block
from https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md which led clients to break if the END_STREAM is received in a DATA frame.

**Comment by deweerdt** [comment]
I believe that this uncovered a bug in the mruby handler where the generator is accessed after calling `h2o_send`. f8faf32 fixes the issue.

**Comment by deweerdt** [comment]
The fuzzer had a crash where `conn` is re-used after `execute_or_enqueue_request` (which can now free `conn` synchronously). In light of that, what do you think of 119ad42 ? That makes fixing mruby unnecessary (because `h2o_send` can't free the generator synchronously, so i've reverted it.)
I think the remaining question is the correctness of the `send_refused_stream` call above. I think it's correct because 1) that code path can't happen with mruby 2) it can't happen with the upgrade path (which the fuzzer hit) either. That said it seems fragile, so it might make sense to make it behave as 119ad42 . Thoughts?

**Comment by kazuho** [comment]
Thanks.

I think you are correct. Looking at our precedents, we have considered disposal of `h2o_req_t` within `h2o_send` as a bug; see https://github.com/h2o/h2o/pull/2602/commits/b27ba5a2cf688889870c4c08c5105d7eccd97a72.

The approach taken in 119ad42 looks consistent to that, and we should fix `send_refused_stream`.

**Comment by deweerdt** [comment]
Thank you @kazuho . daae50c as the fix for `send_refused_stream`

**Comment by kazuho** [comment]
Thank you for the changes. Merged!

**Pull request merged**